src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionTextCache.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq.Expressions;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class ExpressionTextCache
+    {
+        public readonly ConcurrentDictionary<LambdaExpression, string> Entries  = 
+                new ConcurrentDictionary<LambdaExpression, string>(LambdaExpressionComparer.Instance);
+
+        private class LambdaExpressionComparer : IEqualityComparer<LambdaExpression>
+        {
+            public static readonly LambdaExpressionComparer Instance = new LambdaExpressionComparer();
+
+            public bool Equals(LambdaExpression lambdaExpression1, LambdaExpression lambdaExpression2)
+            {
+                // We will cache only pure member access expressions. Hence we compare two expressions 
+                // to be equal only if they are identical member access expressions.
+                var expression1 = lambdaExpression1.Body;
+                var expression2 = lambdaExpression2.Body;
+
+                while (expression1 != null && expression2 != null)
+                {
+                    if (!expression1.EqualsTo(expression2))
+                    {
+                        return false;
+                    }
+
+                    if (expression1.NodeType == ExpressionType.MemberAccess)
+                    {
+                        var memberExpression1 = (MemberExpression)expression1;
+                        expression1 = memberExpression1.Expression;
+
+                        var memberExpression2 = (MemberExpression)expression2;
+                        expression2 = memberExpression2.Expression;
+                    }
+                    else if (expression1.NodeType == ExpressionType.Parameter)
+                    {
+                        expression1 = expression2 = null;
+                    }
+                    else
+                    {
+                        // should never come here.
+                        return false;
+                    }
+                }
+
+                return true;
+            }
```

Comment:
This comparison does not match the hash code implementation and looks generally scary:
- Ignores `Member.Name` of a `MemberExpression`.
- Won't handle `m => m` correctly because it assumes the top-level node is a `MemberExpression`.
- Performs a reference equality comparison on an `Expression` and then checks parts of the expression.
- Does not check whether the `NodeType`s of the two expressions match.

The last two bullets could be fixed by changing [line 30](https://github.com/aspnet/Mvc/pull/4298/files#diff-cbc498e6135dcff3a95d99f007858dd4R30) to
``` c#
if (expression1.NodeType != expression2.NodeType)
```